### PR TITLE
Add the first of the SSE event broadcasting functionality as originally planned

### PR DESCRIPTION
**Pull request by lili7h**
Way back yonder when things were a lot simpler, we created the footings for SSE (Server-Sent Events) broadcasting, with the intention that this channel would be how the backend alerts the frontend to things, without the frontend have to poll for it. We uh, never used it.... UNTIL NOW 

This is a simple implementation, but it lays a lot of ground work that would make copy pasting further SSE broadcasting quite the easier process. 

We broadcast 3 subtypes of ConsoleOutput messages - ChatMessage events, Kill events, and DemoStop events. These are broadcast to subscribers as a JSON blob that looks something like the follow:

```json
{
  "type":"ChatMessage",
  "uuid":"c7bcd50a-6f5d-4a84-8fb7-01b906ed15b7",
  "time":"2024-04-29T11:59:56.979901203Z",
  "event": {
    "player_name":"Lilith",
    "steamid":"76561198071482715",
    "message":"this is for the haters"
  }
}

{
  "type":"PlayerKill",
  "uuid":"c486298d-155b-4ccc-9bbd-876e9e4802ac",
  "time":"2024-04-29T12:00:09.269070783Z",
  "event": {
    "killer_name":"WITCH",
    "killer_steamid":null,
    "victim_name":"Lilith",
    "victim_steamid":"76561198071482715",
    "weapon":"tf_projectile_rocket",
    "crit":false
  }
}

{
  "type":"VoteStarted",
  "content": {
    "count":2,
    "option_1":"ctf_doublecross",
    "option_2":"Extend current Map",
    "option_3":"",
    "option_4":"",
    "option_5":"",
    "voteidx":3
  },
  "time":"2024-04-30T09:37:06.325924920Z",
  "uuid":"f8b59643-1724-4f61-98c6-f59d54d70cb1"
}

{
  "type":"VoteCast",
  "content": {
    "voter":76561198071482715,
    "voter_name":"Lilith",
    "choice":"ctf_doublecross",
    "event": {
      "vote_option":0,
      "team":0,
      "entity_id":1,
      "voteidx":3
    }
  },
  "time":"2024-04-30T09:37:06.325977329Z",
  "uuid":"2f2123ff-b274-4b21-a6c0-6b2f710cd80f"
}
``` 

Note that in the above PlayerKill example, WITCH is a bot, and thus doesnt have a steamID (so null is perfectly expected).

The intention is to use this same framework to broadcast votes to the frontend so it can display votes, vote options, and who voted for what, etc etc. 

Use cases for the current messages? Cute little pop ups in the corner or something that show these things, to display game state and also useful as a meter for whether or not the client is actually properly reading TF2 data. Also, its COOL.

The code is a bit messy, so please eat me alive. Thanks.

## FOR TESTING

You can use the python package `requests-sse` to subscribe to SSE endpoints and display messages as they come in. I have a short demo module (that is literally copy pasted sample code from the PyPI page for requests-sse, wrapped in a poetry project) that you can use: https://github.com/lili7h/http-sse-event-viewer

### CODE REVIEW

**Comment by lili7h** [comment]
ISSUES: Windows build fails due to: `winapi::ctypes::c_void` and `std::ffi::c_void` have similar names, but are actually distinct types

File: src/sse_events.rs
Comment (Bash-09) [inline src/sse_events.rs:55]:
```rust
let event_json = if let Some(demo_msg) = try_get::<DemoMessage>(message) {
    self.handle_demo_message(state, dmo_msg)
} else if let Some(con_msg) = try_get::<ConsoleOutput>(message) {
    self.handle_console_message(state, con_msg)
} else {
    None
};
```

File: src/sse_events.rs
Comment (Bash-09) [inline src/sse_events.rs:65]: I'd personally opt for 
```rust
    ...
}

None
```
But doesn't really matter.

File: src/main.rs
Comment (Bash-09) [inline src/main.rs:223]: Don't forget to remove the commented out stuff if it's gone for good.

File: src/main.rs
Comment (Bash-09) [inline src/main.rs:88]: As above

File: src/main.rs
Comment (Bash-09) [inline src/main.rs:98]: As above

File: src/console.rs
Comment (Bash-09) [inline src/console.rs:64]: Yeah I don't love it either. Hopefully we can find a nicer way to do this at some point.

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:470]: Since it's just used in the broadcast handler thingy and doesn't really need any specific handling of it's internal fields, I'd probably just make the fields public and manipulate them directly instead of using setters.

File: src/sse_events.rs
Comment (Bash-09) [inline src/sse_events.rs:13]: The name `BroadcastableEvent` makes it sound like this struct is the event itself, I liked the naming scheme you had before like `SseBroadcaster` or something along those lines.

File: src/sse_events.rs
Comment (Bash-09) [inline src/sse_events.rs:142]: I know this was just copied from the `PrintVotes` handler, but I think for the SSE messages it would probably be a good idea to provide better defaults, I noticed the struct you use has `Option`s, so maybe they should just be `None` in the fail case.

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:442]: I'd probably make this type non-pub and place it in the sse-events module where it's used, since it doesn't have any utility outside of there.

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:496]: Same as `VoteCastEventWrapped` above.

File: src/console.rs
Comment (Bash-09) [inline src/console.rs:107]: Same as other 2

File: src/console.rs
Comment (Bash-09) [inline src/console.rs:65]: And probably also this

**Review by Bash-09 — COMMENTED** [review]
This is pretty cool so far, I hope we can get something using it soon.

File: src/sse_events.rs
Comment (lili7h) [reply src/sse_events.rs:65]: in that case, i'd have to make the statement inside the if body be a `return`; 
```rs
        if let Some(json) = event_json {
            Handled::<OM>::future(async move {
                broadcast_event(json).await;
                None
            })
        } else {
            None
        }
```
vs
```rs
        if let Some(json) = event_json {
            return Handled::<OM>::future(async move {
                broadcast_event(json).await;
                None
            });
        }
        None
```

File: src/sse_events.rs
Comment (lili7h) [reply src/sse_events.rs:55]: +1

File: src/main.rs
Comment (lili7h) [reply src/main.rs:223]: :clown_face:

File: src/demo.rs
Comment (lili7h) [reply src/demo.rs:470]: +1

File: src/sse_events.rs
Comment (lili7h) [reply src/sse_events.rs:13]: Valid, i had double assigned the `Event` suffix in my mind to two different meanings. Went for `SseEventBroadcaster`.

File: src/sse_events.rs
Comment (Bash-09) [reply src/sse_events.rs:65]: Yeah true. I'm not really bothered either way, was just what I would have done. Up to you.

File: src/sse_events.rs
Comment (lili7h) [reply src/sse_events.rs:142]: +1

File: src/demo.rs
Comment (lili7h) [reply src/demo.rs:442]: well its used in implementing the SerializableVoteEventContent trait,

`impl SerializableVoteEventContent for VoteCastEventWrapped {`

Comment (lili7h) [reply src/demo.rs:442]: line 471 or so

File: src/console.rs
Comment (lili7h) [reply src/console.rs:65]: Ah, i see. You want ALL the traits to to just be in the sse-events. Okay lets see

File: src/console.rs
Comment (lili7h) [reply src/console.rs]: remove the unnessescary serialise fn

File: src/console.rs
Comment (Bash-09) [reply src/console.rs]: Yeah, since the stuff added in this branch isn't really used in demo management, it's all infrastructure for the sse broadcasting stuff, I think it should all go together in there. Rather than bloating this file with stuff that's only used for a specific purpose that's unrelated.

File: src/web.rs
Comment (Bash-09) [inline src/web.rs:728]: :frowning_face:

**Review by Bash-09 — COMMENTED** [review]
Otherwise everything looks pretty good. I'll be happy to approve and merge once you fix up the conflicts and address the last little bit of feedback.

File: src/demo.rs
Comment (lili7h) [reply src/demo.rs:627]: Ideally this vote event does exist in the demo, and does result in events. I'm assuming the parser just fails to pull them out at this point, but I would like the SSE environment be ready to handle these events should they occur. I don't agree with not having a parsing environment set up for these events just because they aren't occuring currently, but should be

File: src/main.rs
Comment (lili7h) [reply src/main.rs:84]: ah ye i see theres kind of an ordering here

File: src/demo.rs
Comment (Bash-09) [reply src/demo.rs:627]: Yes ideally we would get it, but I'm not convinced it's there as the parser seems to parse out all the packets without running into any issues and there could have been reasons for the devs to have changed how votes are initiated. I personally still don't see any reason to add handling for it unless you're totally convinced we will one day see that event, which I'm not, but I'm not going to force you to take it out or anything.

If you want to keep it then please rename the enum variants to reflect their events accurately, add a comment explaining that we should not rely on this event because we've never seen it, and also add that tracing info log back so that if we ever *do* see it then we will know.

Comment (lili7h) [reply src/demo.rs]: Ok. See [27f4890](https://github.com/MegaAntiCheat/client-backend/pull/144/commits/27f48907c92e939ef692e585d7b72b91f4ff9087)

Comment (Bash-09) [reply src/demo.rs]: Cool, I'm happy with that then.

**Comment by Bash-09** [comment]
Looks like you just need to run `cargo fmt` I think

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:781]: We shouldn't be printing this as info, I'm assuming this was intended for debugging to see if it actually worked? But we'd want to remove it or change it to debug level before merging.

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:764]: Wacky syntax, can't you just `for ... in self.shunted_messages.clone().into_iter().enumerate()`? Although it looks like you don't even need an owned copy of this vector so really could it just be `for ... in self.shunted_messages.iter().enumerate()` and get borrows?

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:807]: As above

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:692]: I'd consider making this a `Vec<(u32, VoteCastEvent, Option<SteamID>)>` instead so that you can ensure they are the only type that will be dumped into the vector and you won't have to pattern match to get the vote event back out again later.

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:704]: I think the name could be slightly more informative of what it actually does, as `process` is quite vague. Maybe something along the lines of `get_vote_cast_message`?

File: src/demo.rs
Comment (Bash-09) [reply src/demo.rs:764]: Actually may possibly be redundant, see later messages.

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:794]: I suspect a lot of this could be simplified with `Vec::retain`, something along the lines of:
```rust
self.shunted_messages.retain(|(voteidx, event, steamid)| {
    let Some(output) = self.get_vote_cast_message(event, steamid, state) else {
        return true;
    }

    tracing::info!("{output}");
    false
});
```

(If that causes a complaint about borrowing self while it's mutably borrowed, you could possibly just swap it with an empty vector temporarily while doing the operation, but hopefully that won't be necessary)
```rust
let mut temp = Vec::new();
std::mem::swap(&mut temp, &mut self.shunted_messages);
// Do your stuff here
std::mem::swap(&mut temp, &mut self.shunted_messages);
```

Comment (lili7h) [reply src/demo.rs:794]: Ah i hadn't thought of a retain to filter like this. Makes sense (still not inclined to thinking of closures as a spot to perform arbitrary side affecting actions (oo IO monad!!)). Needed to implement the swap because it doesn't like borrowing self immutably inside the closure for the get_vote_cast_message function call.

File: src/demo.rs
Comment (lili7h) [reply src/demo.rs:704]: ye :+1:

File: src/demo.rs
Comment (lili7h) [reply src/demo.rs:692]: eh i had it like this because i wanted solution that could be used for any message we might want to shunt, but eh ive just renamed it to `shunted_vote_cast_message` and done as suggested

Comment (Bash-09) [reply src/demo.rs]: Ah yeah that is a reasonable justification, although I think for the purpose of this handler it probably won't need to do this with any other ones in the future, and if it does it should be a pretty easy change to make.

File: src/demo.rs
Comment (Bash-09) [reply src/demo.rs]: > Needed to implement the swap because it doesn't like borrowing self immutably inside the closure for the get_vote_cast_message function call.

Ah well

> still not inclined to thinking of closures as a spot to perform arbitrary side affecting actions (oo IO monad!!)

Yeah I totally getcha, it does feel weird! And probably wouldn't be ideal for everything. I think in a similar case I would opt to pull the ones that need to be dealt with out with something like `Vec::extract_if` (formerly `Vec::drain_filter`), but that isn't stabilised yet so requires nightly, and the fact it changed from `drain_filter` to `extract_if` is an example of nightly code having breaking changes between versions (this change did stop a previous project of mine from compiling when I updated my compiler).

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:773]: I don't think this if statement is even necessary is it? Considering `get_vote_cast_message` will return `None` if this were false anyways, and I don't think it's a particularly expensive function to call (especially at the frequency it would be called anyways (that is to say, not often)).

In a similar vein, the check above to break out early isn't strictly necessary either as the vector will just be empty so nothing happens.

Just a minor preference thing you can change if you want, but not a huge deal either way.

**Review by Bash-09 — APPROVED** [review]

File: src/demo.rs
Comment (lili7h) [reply src/demo.rs]: removing that if lets me remove a field from the shunted message tuple, so i did it. Leaving the other one that saves a mem swap

File: src/demo.rs
Comment (Bash-09) [inline src/demo.rs:761]: Saves?

**Review by Bash-09 — APPROVED** [review]
Assuming from your testing it still all behaves as expected. I haven't had a chance to do any in-depth testing but I trust your judgement and as far as I can tell the code looks fine :thumbsup:

**Pull request merged**
